File: DSA/NiharikaChand_590017697/Day26Question2.java
import java.util.*;
class Day26Question2 
{
    static class Node 
    {
        int data;
        Node next;
        Node(int data) 
        {
            this.data = data;
            this.next = null;
        }
    }
    Node head;

    public void insert(int data) 
    {
        Node newNode = new Node(data);
        if (head == null) 
        {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) 
        {
            temp = temp.next;
        }
        temp.next = newNode;
    }

    public Node union(Node head1, Node head2) 
    {
        Node cHead = null, cTail = null;
        Node temp = head1;
        while (temp != null) 
        {
            Node newNode = new Node(temp.data);
            if (cHead == null) 
            {
                cHead = newNode;
                cTail = newNode;
            } 
            else 
            {
                cTail.next = newNode;
                cTail = newNode;
            }
            temp = temp.next;
        }

        temp = head2;
        while (temp != null) 
        {
            Node newNode = new Node(temp.data);
            cTail.next = newNode;
            cTail = newNode;
            temp = temp.next;
        }

        cHead = sortList(cHead);
        cHead = remDup(cHead);
        return cHead;
    }

    private Node sortList(Node head) 
    {
        if (head == null || head.next == null)
            return head;

        Node i, j;
        for (i = head; i.next != null; i = i.next) 
        {
            for (j = i.next; j != null; j = j.next) 
            {
                if (i.data > j.data) 
                {
                    int temp = i.data;
                    i.data = j.data;
                    j.data = temp;
                }
            }
        }
        return head;
    }

    private Node remDup(Node head) 
    {
        Node current = head;
        while (current != null && current.next != null) 
        {
            if (current.data == current.next.data) 
            {
                current.next = current.next.next; 
            } 
            else 
            {
                current = current.next;
            }
        }
        return head;
    }

    public void display(Node head) 
    {
        Node temp = head;
        while (temp != null) 
        {
            System.out.print(temp.data + "->");
            temp = temp.next;
        }
        System.out.println("NULL");
    }

    public static void main(String args[]) 
    {
        Scanner sc = new Scanner(System.in);
        Day26Question2 list1 = new Day26Question2();
        Day26Question2 list2 = new Day26Question2();
        System.out.print("Enter size of linkedlist1: ");
        int n1 = sc.nextInt();
        for (int i = 0; i < n1; i++) 
        {
            System.out.print("Enter element for L1: ");
            int data = sc.nextInt();
            list1.insert(data);
        }

        System.out.print("Enter size of linkedlist2: ");
        int n2 = sc.nextInt();
        for (int i = 0; i < n2; i++) 
        {
            System.out.print("Enter element for L2: ");
            int data = sc.nextInt();
            list2.insert(data);
        }

        Day26Question2 obj = new Day26Question2();
        Node unionHead = obj.union(list1.head, list2.head);

        System.out.println("Output:");
        obj.display(unionHead);
    }
}

File: DSA/Gautam_Shyamnani_590013078/Day26Question2.cpp
#include <iostream>
#include <set>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node(int val) : data(val), next(nullptr) {}
};


Node* makeUnion(Node* head1, Node* head2) {
    set<int> values;

    
    Node* temp = head1;
    while (temp) {
        values.insert(temp->data);
        temp = temp->next;
    }


    temp = head2;
    while (temp) {
        values.insert(temp->data);
        temp = temp->next;
    }


    Node* newHead = nullptr;
    Node* tail = nullptr;
    for (int val : values) {
        Node* newNode = new Node(val);
        if (!newHead) {
            newHead = newNode;
            tail = newNode;
        } else {
            tail->next = newNode;
            tail = tail->next;
        }
    }

    return newHead;
}

void printList(Node* head) {
    while (head) {
        cout << head->data << " -> ";
        head = head->next;
    }
    cout << "nullptr\n";
}


int main() {

    Node* L1 = new Node(9);
    L1->next = new Node(6);
    L1->next->next = new Node(4);
    L1->next->next->next = new Node(2);
    L1->next->next->next->next = new Node(3);
    L1->next->next->next->next->next = new Node(8);


    Node* L2 = new Node(1);
    L2->next = new Node(2);
    L2->next->next = new Node(8);
    L2->next->next->next = new Node(6);
    L2->next->next->next->next = new Node(2);

    Node* unionList = makeUnion(L1, L2);
    cout << "Union Linked List:\n";
    printList(unionList);
}
